File: app/modules/omx/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel

OmxAction = Literal[
    "version",
    "status",
    "doctor",
    "doctor_team",
    "cleanup",
    "cleanup_dry_run",
    "cancel",
    "reasoning_get",
    "reasoning_set",
]
OmxReasoningLevel = Literal["low", "medium", "high", "xhigh"]


class OmxOverviewResponse(DashboardModel):
    available: bool
    binary_path: str | None = None
    runtime_env_path: str | None = None
    runtime_dir: str | None = None
    version: str | None = None
    reasoning: OmxReasoningLevel | None = None
    status_summary: str | None = None
    doctor_summary: str | None = None
    warnings: list[str] = Field(default_factory=list)
    last_checked_at: datetime


class OmxQuickRef(DashboardModel):
    key: str
    label: str
    command: str
    description: str


class OmxDashboardSession(DashboardModel):
    id: str
    status: str
    context_line: str
    started_at: datetime | None = None
    last_activity_at: datetime | None = None
    cwd: str | None = None
    source: str


class OmxDashboardWorker(DashboardModel):
    team: str
    worker_id: str
    status: str
    job_line: str
    role: str | None = None
    last_heartbeat_at: datetime | None = None
    session_id: str | None = None


class OmxSessionTokenUsage(DashboardModel):
    session_id: str
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    exact: bool = False


class OmxWorkerTokenUsage(DashboardModel):
    team: str
    worker_id: str
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    exact: bool = False


class OmxTokenUsage(DashboardModel):
    sessions: list[OmxSessionTokenUsage] = Field(default_factory=list)
    workers: list[OmxWorkerTokenUsage] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class OmxDashboardResponse(DashboardModel):
    quick_refs: list[OmxQuickRef] = Field(default_factory=list)
    sessions: list[OmxDashboardSession] = Field(default_factory=list)
    workers: list[OmxDashboardWorker] = Field(default_factory=list)
    token_usage: OmxTokenUsage = Field(default_factory=OmxTokenUsage)
    warnings: list[str] = Field(default_factory=list)
    updated_at: datetime


class OmxRunRequest(DashboardModel):
    action: OmxAction
    level: OmxReasoningLevel | None = None


class OmxCommandResultResponse(DashboardModel):
    action: OmxAction
    command: list[str]
    exit_code: int
    stdout: str
    stderr: str
    started_at: datetime
    finished_at: datetime
    timed_out: bool = False

File: frontend/src/features/omx/schemas.ts
import { z } from "zod";

export const OmxReasoningLevelSchema = z.enum(["low", "medium", "high", "xhigh"]);

export const OmxOverviewSchema = z.object({
  available: z.boolean(),
  binaryPath: z.string().nullable(),
  runtimeEnvPath: z.string().nullable(),
  runtimeDir: z.string().nullable(),
  version: z.string().nullable(),
  reasoning: OmxReasoningLevelSchema.nullable(),
  statusSummary: z.string().nullable(),
  doctorSummary: z.string().nullable(),
  warnings: z.array(z.string()),
  lastCheckedAt: z.string().datetime({ offset: true }),
});

export const OmxQuickRefSchema = z.object({
  key: z.string(),
  label: z.string(),
  command: z.string(),
  description: z.string(),
});

export const OmxDashboardSessionSchema = z.object({
  id: z.string(),
  status: z.string(),
  contextLine: z.string(),
  startedAt: z.string().datetime({ offset: true }).nullable().optional().default(null),
  lastActivityAt: z.string().datetime({ offset: true }).nullable().optional().default(null),
  cwd: z.string().nullable().optional().default(null),
  source: z.string(),
});

export const OmxDashboardWorkerSchema = z.object({
  team: z.string(),
  workerId: z.string(),
  status: z.string(),
  jobLine: z.string(),
  role: z.string().nullable().optional().default(null),
  lastHeartbeatAt: z.string().datetime({ offset: true }).nullable().optional().default(null),
  sessionId: z.string().nullable().optional().default(null),
});

export const OmxSessionTokenUsageSchema = z.object({
  sessionId: z.string(),
  inputTokens: z.number().int().nullable().optional().default(null),
  outputTokens: z.number().int().nullable().optional().default(null),
  totalTokens: z.number().int().nullable().optional().default(null),
  exact: z.boolean(),
});

export const OmxWorkerTokenUsageSchema = z.object({
  team: z.string(),
  workerId: z.string(),
  inputTokens: z.number().int().nullable().optional().default(null),
  outputTokens: z.number().int().nullable().optional().default(null),
  totalTokens: z.number().int().nullable().optional().default(null),
  exact: z.boolean(),
});

export const OmxTokenUsageSchema = z.object({
  sessions: z.array(OmxSessionTokenUsageSchema),
  workers: z.array(OmxWorkerTokenUsageSchema),
  notes: z.array(z.string()),
});

export const OmxDashboardSchema = z.object({
  quickRefs: z.array(OmxQuickRefSchema),
  sessions: z.array(OmxDashboardSessionSchema),
  workers: z.array(OmxDashboardWorkerSchema),
  tokenUsage: OmxTokenUsageSchema,
  warnings: z.array(z.string()),
  updatedAt: z.string().datetime({ offset: true }),
});

export const OmxActionSchema = z.enum([
  "version",
  "status",
  "doctor",
  "doctor_team",
  "cleanup",
  "cleanup_dry_run",
  "cancel",
  "reasoning_get",
  "reasoning_set",
]);

export const OmxRunCommandRequestSchema = z.object({
  action: OmxActionSchema,
  level: OmxReasoningLevelSchema.optional(),
});

export const OmxRunCommandResultSchema = z.object({
  action: OmxActionSchema,
  command: z.array(z.string()),
  exitCode: z.number().int(),
  stdout: z.string(),
  stderr: z.string(),
  startedAt: z.string().datetime({ offset: true }),
  finishedAt: z.string().datetime({ offset: true }),
  timedOut: z.boolean(),
});

export type OmxOverview = z.infer<typeof OmxOverviewSchema>;
export type OmxDashboard = z.infer<typeof OmxDashboardSchema>;
export type OmxAction = z.infer<typeof OmxActionSchema>;
export type OmxReasoningLevel = z.infer<typeof OmxReasoningLevelSchema>;
export type OmxRunCommandRequest = z.infer<typeof OmxRunCommandRequestSchema>;
export type OmxRunCommandResult = z.infer<typeof OmxRunCommandResultSchema>;
